### PR TITLE
Simple REDIRECT

### PR DESCRIPTION
**Pull request by ksdenisova**
**Feature**: Simple Redirect

  **Scenario**: Getting a resource that has been moved to a different location
    **Given** I make an GET request to "/redirect"
    **Then** my response should have status code 301
    **And** my response should have a location header with the "http://0.0.0.0:5000/simple_get" URI
    **And** my response should have an empty body

[Trello](https://trello.com/c/ZGB947mr/14-simple-redirect)

image

**Refactoring:**
- Add `status_code_reason` to `Response`
- Update `Routes` to use function instead of map
- Add module name before `Response` and `Request`
- Add Parameterized tests support
- Add `Headers` struct

### CODE REVIEW

File: lib/http_server/header.ex [hunk 1]
```diff
@@ -0,0 +1,5 @@
+defmodule HttpServer.Header do
```
Comment (Aoxian) [inline lib/http_server/header.ex:1]: why is this file needed?  I find it troubling that a redirect location is hard coded into http_server library code, when it is part of an acceptance test.

File: lib/http_server/header.ex [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+  @location "Location: http://127.0.0.1:5000/simple_get"
+
+  def location, do: @location
+end
```
Comment (Aoxian) [inline lib/http_server/header.ex:5]: why is this file needed?  I find it troubling that a redirect location is hard coded into http_server library code, when it is part of an acceptance test.

Comment (ksdenisova) [reply lib/http_server/header.ex:5]: Using this file I wanted to have different headers as constant values. But location is wrong. It might be better if I remove it and specify the location as a String:
```
def handle(_request, builder) do
    builder.new_response()
    |> builder.add_status_code(301)
    |> builder.add_headers(["Location: http://127.0.0.1:5000/simple_get"])
end
```

Comment (Aoxian) [reply lib/http_server/header.ex:5]: I like your thinking, as there are many different headers to eventually support. What would be a better way to have a place for default headers, that you can easily override, and then systematically attach to a response?

File: lib/http_server_fixture/handler/simple_redirect.ex [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule HttpServerFixture.Handler.SimpleRedirect do
+  @behaviour HttpServer.Handler
+
+  @impl HttpServer.Handler
+  def handle(_request, builder) do
+    builder.new_response()
+    |> builder.add_status_code(301)
```
Comment (Aoxian) [inline lib/http_server_fixture/handler/simple_redirect.ex:7]: Now that you are adding [Status Code 301](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/301), is the response you are returning including all the correct information?

Expected:
```shell
HTTP/1.1 301 Moved Permanently
Location: http://127.0.0.1:5000/simple_get
```

Comment (Aoxian) [reply lib/http_server_fixture/handler/simple_redirect.ex:7]: To be fair, the acceptance test is not requiring it but maybe its just an in complete acceptance test.

File: lib/routes.ex [hunk 4]
```diff
@@ -4,6 +4,7 @@ defmodule Routes do
   def build_routes_map do
     %{"/echo_body" => %{"POST" => Handler.SimplePost},
       "/simple_get" => %{"GET" => Handler.SimpleGet},
-      "/simple_get_with_body" => %{"GET" => Handler.SimpleGetWithBody}}
+      "/simple_get_with_body" => %{"GET" => Handler.SimpleGetWithBody},
+      "/redirect" => %{"GET" => Handler.SimpleRedirect}}
```
Comment (Aoxian) [inline lib/routes.ex:8]: There appears to be a pattern here where a `path` is mapped to another mapping of `method` to `handler`.  Can something be done to enforce this pattern?  What is stopping someone from adding something like this?

```elixir
%{"this is not a path" =? %{"looky here!" => MyStrange.Module, your_friend: "Ian"}}
```
```

Comment (Aoxian) [reply lib/routes.ex:8]: Perhaps you could use Elixirs built in pattern matching to your advantage in some way?

File: lib/http_server_fixture/handler/simple_redirect.ex (hunk 3 above)
Comment (ksdenisova) [reply lib/http_server_fixture/handler/simple_redirect.ex:7]: I've checked that tests still pass if we change status code from `301` to `301 Moved Permanently`. I see that in acceptance test they only checks the code number:
```
step 'my response should have status code 301' do
      expect(@response.status_code).to eq 301
end
```

Comment (Aoxian) [reply lib/http_server_fixture/handler/simple_redirect.ex:7]: 301 is the `status_code` and "Moved Permanently" would be the `status_code_reason`.  Can we 'update' the acceptance test? then solve the problem still allowing a user to only add the `status_code`? These codes and reasons are well defined in the HTTP spec after all, and should not be changing anytime soon.

Comment (ksdenisova) [reply lib/http_server_fixture/handler/simple_redirect.ex:7]: I understand that this is a format of HTTP Request and it is not going to change anytime soon. If we want to change the acceptance tests, we need to change the `Response` class, then add the step with checking `status_code_reason` to each acceptance test, and check if it parses correctly. What if instead of this we add the `status_code_reason` to our `Response` and write the Unit tests?

File: lib/routes.ex (hunk 4 above)
Comment (ksdenisova) [reply lib/routes.ex:8]: Can we pair on this today?

File: lib/http_server/header.ex (hunk 2 above)
Comment (ksdenisova) [reply lib/http_server/header.ex:5]: What if we add functions to the `Response.Builder` module to add different headers? I also see that location URL can be different, I don't need this specific location as a constant. There is example:
```
def add_location_header(response, location) do
    ...
end
```

Comment (Aoxian) [reply lib/http_server/header.ex:5]: That's the OO answer, what's the functional answer?

Comment (Aoxian) [reply lib/http_server/header.ex:5]: Imagine you will eventually support all the headers in the RFC.  Here is a (very long list) to help you think: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

Comment (ksdenisova) [reply lib/http_server/header.ex]: Added `Headers` struct.

File: lib/http_server_fixture/handler/simple_redirect.ex (hunk 3 above)
Comment (ksdenisova) [reply lib/http_server_fixture/handler/simple_redirect.ex:11]: Added `status_code_reason` to `Response`.

File: lib/routes.ex (hunk 4 above)
Comment (ksdenisova) [reply lib/routes.ex]: Changed map to functions with pattern matching.

File: lib/http_server/response/headers.ex [hunk 5]
```diff
@@ -0,0 +1,10 @@
+defmodule HttpServer.Response.Headers do
+  defstruct [:Location]
+
+  def defined_headers(headers) do
+    Map.from_struct(headers)
+    |> Map.filter(fn {_key, value} -> value != nil end)
+    |> Map.to_list()
+    |> Enum.map(fn {key, value} -> Enum.join([key, value], ": ") end)
```
Comment (Aoxian) [inline lib/http_server/response/headers.ex:8]: 👌 looks great!

File: lib/http_server/router.ex [hunk 6]
```diff
@@ -14,25 +13,7 @@ defmodule HttpServer.Router do
...
-    end
-  end
-
-  defp get_handler(route, method) do
-    case Map.get(route, method) do
-      nil ->
-        raise Exception.MethodNotFoundError
-      _ = handler -> handler
-    end
+    routes.handle(request)
```
Comment (Aoxian) [inline lib/http_server/router.ex:16]: this reads like a sentence! Nice!

**Review by Aoxian — APPROVED** [review]

File: lib/routes.ex (hunk 4 above)
Comment (Aoxian) [reply lib/routes.ex]: Nice work

File: lib/http_server/router.ex (hunk 6 above)
Comment (ksdenisova) [reply lib/http_server/router.ex:16]: Thank you!
